python/lsst/analysis/drp/gatherResourceUsage.py

```diff
@@ -0,0 +1,550 @@
+        parser.add_argument(
+            "collections",
+            type=str, nargs="+",
+            help="Collection(s)s to search for input metadata.",
+        )
+        parser.add_argument(
+            "--dataset-types",
+            type=str,
+            action="extend",
+            help=(
+                "Glob-style patterns for input metadata dataset types.  If a pattern matches a "
+                "non-metadata dataset type, non-metadata matches will be ignored."
+            ),
+        )
+        parser.add_argument(
+            "--where",
+            type=str,
+            default=None,
+            help="Data ID expression used when querying for input metadata datasets.",
+        )
+        args = parser.parse_args(argv)
+        base_dataset_type_filter = re.compile(r"\w+_metadata")
+        if not args.dataset_types:
+            input_dataset_types = base_dataset_type_filter
+        else:
+            input_dataset_types = [fnmatch.translate(expr) for expr in args.dataset_types]
+        butler = Butler(args.repo, collections=args.collections)
+        metadata_refs = butler.registry.queryDatasets(input_dataset_types, where=args.where, findFirst=True)
+        qg = cls.build_quantum_graph(metadata_refs)
+        qg.saveUri(args.filename)
+
+    def runQuantum(
+        self,
+        butlerQC,
+        inputRefs,
+        outputRefs,
+    ):
+        # Docstring inherited.
+        # This override exists just so we can pass the butler registry's
+        # DimensionUniverse to run in order to standardize the dimensions.
+        inputs = butlerQC.get(inputRefs)
+        outputs = self.run(butlerQC.registry.dimensions, **inputs)
+        butlerQC.put(outputs, outputRefs)
+
+    def run(self, universe, input_metadata):
+        """Gather resource usage statistics from per-quantum metadata.
+
+        Parameters
+        ----------
+        universe : `DimensionUniverse`
+            Object managing all dimensions recognized by the butler; used to
+            standardize and expand `GatherResourceUsageConfig.dimensions`.
+        input_metadata : `list` [ `DeferredDatasetHandle` ]
+            List of `lsst.daf.butler.DeferredDatasetHandle` that can be used to
+            load all input metadata datasets.
+
+        Returns
+        -------
+        result : `Struct`
+            Structure with a single element:
+
+            - ``outout_table``: a `pandas.DataFrame` that aggregates the
+              configured resource usage statistics.
+        """
+        dimensions = universe.extract(self.config.dimensions)
+        # Transform input list into a dict keyed by data ID.
+        handles_by_data_id = {}
+        for handle in input_metadata:
+            handles_by_data_id[handle.dataId] = handle
+        n_rows = len(handles_by_data_id)
+        # Create a dict of empty column arrays that we'll ultimately make into
+        # a table.
+        columns = {
```

Comment:
why use this dict of numpy arrays vs a recarray

Reply:
`pandas.DataFrame` is a transpose of `recarray` but pretty much identical to a dict of numpy arrays under the hood.  It may still do a copy when I convert it to a `DataFrame`, but at least it won't need to do a transpose.